fix: handle Event signals gracefully

subpop opened the pull request:
Return errors when handling Event signals emitted by workers and handle
the optional 'message' parameter gracefully.

Signed-off-by: Link Dupont <link@sub-pop.net>

jirihnidek commented on internal/work/errors.go:
```diff
@@ -0,0 +1,39 @@
+
+import "reflect"
+
+var errIncorrectSignal workError = newWorkError("incorrect signal")
```
Type `workError` can be omitted.

jirihnidek commented on internal/work/errors.go:
```diff
@@ -0,0 +1,39 @@
+import "reflect"
+
+var errIncorrectSignal workError = newWorkError("incorrect signal")
+var errMissingRequiredParameters workError = newWorkError("missing required parameters")
```
Same comment here.

jirihnidek commented on internal/work/dispatcher.go:
```diff
@@ -340,3 +321,42 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
+// workerEventFromSignal creates an ipc.WorkerEvent from a DBus signal.
+func workerEventFromSignal(s *dbus.Signal) (*ipc.WorkerEvent, error) {
+	if s.Name != "com.redhat.Yggdrasil1.Worker1.Event" {
+		return nil, errIncorrectSignal
```
We are loosing information here, what signal is wrong and what signal is expected. We only return error message: `incorrect signal`. This could be improved.

jirihnidek commented on internal/work/dispatcher.go:
```diff
@@ -340,3 +321,42 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
+	}
+
+	if len(s.Body) < 2 {
+		return nil, errMissingRequiredParameters
```
We could improve error message here and tell user/developer what parameters are required.

jirihnidek commented on internal/work/dispatcher.go:
```diff
@@ -340,3 +321,42 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
+
+	event := ipc.WorkerEvent{}
+
+	if len(s.Body) > 0 {
```
I think that we can remove these checks for the length of `s.Body`, because we have already checked it

subpop replied:
I was trying to find a pattern that is recognizable and that clearly indicates which indices of the body slice will be accessed. I know the checks are superfluous, but they serve as sentinels to indicate that certain length slices are going to be used. I might be able to rework this into a switch using fallthrough cases, but I'm not convinced that would more clearly express my intention.

jirihnidek replied:
Hint: you can simply use comments in code ;-).

jirihnidek commented on internal/work/dispatcher.go:
```diff
@@ -340,3 +321,42 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
+		event.Name = ipc.WorkerEventName(name)
+	}
+
+	if len(s.Body) > 1 {
```
Same comment here.

jirihnidek commented on internal/work/dispatcher.go:
```diff
@@ -340,3 +321,42 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
+		event.MessageID = messageID
+	}
+
+	if len(s.Body) > 2 {
```
And here.

jirihnidek reviewed (changes requested):
Overall, it looks good. :+1: I have few comments and suggestions.

subpop commented:
I took a different approach towards unpacking a signal into a WorkerEvent. I dropped the two global variables; they weren't useful without adding variability to them, and after that happens, there is no reason to keep global error variables at all. I renamed the unpack error into a more descriptively named "typeConversionError", since that's more accurately describing what the error carries. I dropped the error checking for signal name and parameter length. Looping over the body slice and attempting to convert the `interface{}` types to their appropriate types will catch any errors that might exist from incorrect signal bodies.

jirihnidek reviewed (approved):
LGTM, Thanks for update
